### PR TITLE
Supporting changes for PR#1012

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

- Const-correctness in MeasureCounts: in case we need to pass a `const MeasureCounts&` (e.g., Json conversion)

- Add string broadcast to MPI interface.

- Minor tidy up some somments/include.

- nvq++ in library mode should also include `OBJS_TO_MERGE` when doing the linking (compiled `backendConfig.cpp`)

### CODE REVIEW

**Comment by 1tnguyen** [comment]
Note: I'm breaking up #1012 into smaller chunks to ease the review process.

File: runtime/cudaq/utils/registry.h [hunk 1]
```diff
@@ -7,8 +7,7 @@
  ******************************************************************************/
 
 #pragma once
-#define LLVM_DISABLE_ABI_BREAKING_CHECKS_ENFORCING 1
-#include "llvm/Support/Registry.h"
```
Comment (schweitzpgi) [inline runtime/cudaq/utils/registry.h:11]: Are you sure we should remove this? LLVM will crash if there is an ABI version mismatch.

**Review by schweitzpgi — COMMENTED** [review]
+1

File: runtime/cudaq/utils/registry.h (hunk 1 above)
Comment (1tnguyen) [reply runtime/cudaq/utils/registry.h:11]: We don't use `llvm/Support/Registry.h` in this header. The other place (`runtime/common/Registry.h`) where we actually use `llvm::Registry` already has that `#define`.

We can compile without this `#include` here, so I guess it should be safe.

File: runtime/cudaq/platform/QuantumExecutionQueue.h [hunk 2]
```diff
@@ -38,8 +38,11 @@ class QuantumExecutionQueue {
   /// Enqueue a Sampling task.
   void enqueue(QuantumTask &task);
 
+  /// Get id of the thread this queue executes on.
+  std::thread::id getExecutionThreadId() const;
+
 protected:
-  /// The mutex, used for locking when adding to the queue
+  /// The `mutex`, used for locking when adding to the queue
```
[the remark this answers is not in the document]
Comment (1tnguyen) [reply runtime/cudaq/platform/QuantumExecutionQueue.h]: Apparently, `//!` didn't work. I've reverted the change and added "mutex" to the dictionary.

File: runtime/common/MeasureCounts.cpp [hunk 3]
```diff
@@ -312,12 +313,16 @@ std::size_t sample_result::size(const std::string_view registerName) noexcept {
...
+                                  const std::string_view registerName) const {
   auto iter = sampleResults.find(registerName.data());
   if (iter == sampleResults.end())
     return 0.0;
 
-  return (double)iter->second.counts[bitStr.data()] / totalShots;
+  const auto countIter = iter->second.counts.find(bitStr.data());
+  if (countIter == iter->second.counts.end())
+    return 0.0;
+  else
```
Comment (schweitzpgi) [inline runtime/common/MeasureCounts.cpp:324]: nit: else after return

File: runtime/cudaq/cudaq.cpp [hunk 4]
```diff
@@ -185,6 +185,11 @@ void broadcast(std::vector<double> &data, int rootRank) {
   commPlugin->broadcast(data, rootRank);
 }
 
+void broadcast(std::string &data, int rootRank) {
```
Comment (schweitzpgi) [inline runtime/cudaq/cudaq.cpp:188]:
```suggestion
void broadcast(const std::string &data, int rootRank) {
```
?

File: runtime/cudaq/utils/registry.h (hunk 1 above)
Comment (amccaskey) [reply runtime/cudaq/utils/registry.h:11]: This is old - from when the registry code was in utils and not in common. Safe to remove

File: runtime/cudaq/distributed/mpi_plugin.cpp [hunk 5]
```diff
@@ -92,6 +92,17 @@ void MPIPlugin::broadcast(std::vector<double> &data, int rootRank) {
       m_comm, data.data(), data.size(), FLOAT_64, rootRank));
 }
 
+void MPIPlugin::broadcast(std::string &data, int rootRank) {
+  std::int32_t strLen = data.size();
+  HANDLE_MPI_ERROR(
+      m_distributedInterface->Bcast(m_comm, &strLen, 1, INT_32, rootRank));
+
+  if (rank() != rootRank)
+    data.resize(strLen);
```
Comment (schweitzpgi) [inline runtime/cudaq/distributed/mpi_plugin.cpp:101]: This seems redundant, but maybe the call to `Bcast` is changing the size of the string object... :shrug:

**Review by amccaskey — APPROVED** [review]
nice fixes / clean up. +1

File: runtime/common/MeasureCounts.cpp (hunk 3 above)
Comment (1tnguyen) [reply runtime/common/MeasureCounts.cpp]: Good catch. I've updated the code to use a ternary operator instead.

File: runtime/cudaq/cudaq.cpp (hunk 4 above)
Comment (1tnguyen) [reply runtime/cudaq/cudaq.cpp:188]: The semantics of these `broadcast` APIs is that other ranks (not `rootRank`) will have its `data` updated; hence cannot be `const`.

File: runtime/cudaq/distributed/mpi_plugin.cpp (hunk 5 above)
Comment (1tnguyen) [reply runtime/cudaq/distributed/mpi_plugin.cpp:101]: Indeed, we need to allocate these string buffers before using MPI API to populate them.
Generally, only `rootRank` knows the size of the string, hence we need to do a two-pass `Bcast` as above.

File: runtime/cudaq/cudaq.cpp (hunk 4 above)
Comment (schweitzpgi) [reply runtime/cudaq/cudaq.cpp:188]: +1

**Pull request merged**